### ISSUE TITLE
Accept recursive calls to templatefile()

### ISSUE THREAD

**Issue by Exchizz** [opentffoundation/opentf#402]
### OpenTF Version

```shell
1
```

### Use Cases

2

### Attempted Solutions

4

### Proposal

5

### References

6

**Comment by eranelbaz** [comment]
Hey @Exchizz 

To help us maintain a clear separation between OpenTF and HashiCorp's offerings, we're asking that people describe issues the issues again other rather than linking those directly.

I've thus scrubbed out any links to said issues/PRs.

If there's any more context or description to the problem you think would be good to share and add in, please do.

Thanks!

**Comment by eranelbaz** [comment]
Can you please share the outcome of referencing the same file?

**Comment by Exchizz** [comment]
My college and I had a chat about how this could be implemented. We suggest having a maximum recursion depth counter that is incremented every time `templatefile()` includes a `templatefile()`. If that counter then exceeds fx. 300, then opentf should throw an error saying the maximum include depth has been reached.

**Comment by maunzCache** [comment]
What about tracking the members that do the call to the `templatefile()` and only forbid to include any parent again? This can protect against recursive includes and would limit the overall depth to some degree.

**Comment by cube2222** [comment]
This makes sense. We won't be prioritizing this, but it's open to community involvement to write up and RFC for this (please include the technical details).

**Comment by cam72cam** [comment]
To be clear, we need to figure out how to pass the stack/depth through the evaluation context and minimize the complexity of doing so.

**Comment by maunzCache** [comment]
> One approach comes to my mind that eliminates a need for a hard cap on the number of recursive calls. A cycle detection algorithm could be run after every, say 100, calls to the function. But this is worth implementing only if no limit on recursion is important because of the overhead of maintaining a graph.

What i think might be the best is to leave it to the user. By default allow unlimited recursion (or none) but let the user control to override it with a hard cap (or none). As this would be a feature that would only work in tofu, i'd need to be supported either via parameters of the cli or environment variables. But this would be an excellent discussion for the RFC i guess :)

**Comment by Yantrio** [comment]
It's also worth noting that we also have `templatestring` arriving soon, and we should try and come up with a solution that works for both.

See https://github.com/opentofu/opentofu/issues/301 for progress.

**Comment by cam72cam** [comment]
Got a bit nerd-sniped this afternoon and took a swing at this: https://github.com/opentofu/opentofu/pull/1250

It's also worth looking into as `templatestring` is in-progress and may interact with this in interesting ways.

**Comment by cam72cam** [comment]
My PR is intended as a pathfinder / discussion point and should not preempt any discussions on how the recursion should be tracked / cycles detected.

**Comment by skyzyx** [comment]
Thinking about performance and memory usage, rather than recursing (which can lead to stack-depth issues), what about approaching this as multi-pass templating approach (queue-based, flat stack)?

Rather than writing directly to the output file like this…

```hcl
resource "local_file" "pipeline-templater" {
  content  = templatefile("pipeline.tftpl", { vars = local.variables })
  filename = "pipeline.yaml"
}
```

…it writes to a temp file first, iteratively applies changes, and then writes the final file to `pipeline.yaml`.

My (very fake, very untested) code under the hood would look something like this:

```go
package main

import (
    "io"
    "os"
)

type Work struct {
    Import string
    Vars   map[string]any
}

func Templatefile(filename string, tmplFile string, tmplVars map[string]any) error {
    // Need to tell it to do the first one first.
    workToDo := []Work{
        {
            Import: tmplFile,
            Vars:   tmplVars,
        },
    }

    // Create a temporary file, but smarter than this.
    tmpF, err := os.Create("/tmp/file")
    if err != nil {
        return err
    }

    // While there's work to be done, do work.
    for len(workToDo) > 0 {
        workToDo = append(workToDo, funcWhichParsesTheFunctionCallsAndWritesToTempFile(tmplFile, tmplVars)...)
    }

    // All done processing. Read the temp file and write the data in an efficient way.
    byts, err := io.ReadAll(tmpF)
    if err != nil {
        return err
    }

    outF, err := os.Create(filename)
    if err != nil {
        return err
    }

    // Write the content to the file.
    _, err = outF.Write(byts)
    if err != nil {
        return err
    }

    // Close the tmp file.
    err = tmpF.Close()
    if err != nil {
        return err
    }

    // Close the output file.
    err = outF.Close()
    if err != nil {
        return err
    }

    return nil
}
```

In short, as you encounter a reference to `templatefile()`, you add it to the queue. You iterate over the queue performing the replacements of `templatefile()` with the resolved data until the queue is empty. Then you write the data to a file to the requested location.

You still run the risk of an infinite loop (maybe also track the number of replacements you've performed and set a limit), but you do it using _far_ less memory.

**Comment by birjj** [comment]
@skyzyx Wouldn't that risk unintentional code execution (e.g. if a variable happens to include a value that is templater syntax it would get templated in the next iteration), which could be hard to anticipate or debug?

I know the LaTeX community has had various problems with the multi-pass rendering approach it uses, as it puts strict requirements on idempotency, and makes the whole rendering process hard to reason about.

**Comment by cam72cam** [comment]
Given how the HCL EvalContext is managed through evaluation, not doing this recursively is a fairly tricky proposition.

I'm going to write up an RFC based on my initial PR and comment below.  My goal is to provide a version that works and proves out the concept.  If limitations are commonly hit with the initial version, changes to the default depth can be made or potential optimizations can be made to the implementation.

**Comment by skyzyx** [comment]
From what appears to be a deleted comment by _Johan Fagerberg_, that's still in my email:

> @skyzyx Wouldn't that risk unintentional code execution (e.g. if a variable happens to include a value that is templater syntax it would get templated in the next iteration), which could be hard to anticipate or debug?

This is a possibility with multi-pass (or nested) replacements. I suspect that HashiCorp limited this to a single replacement for simplicity. If OpenTofu were to add support for nested templatefiles, it would be wise to have comprehensive tests and to also document gotchas in the public-facing documentation.

But the feature _itself_ is complex.

---

@cam72cam said:

> Given how the HCL EvalContext is managed through evaluation, not doing this recursively is a fairly tricky proposition.

I'll have to take your word for it. I just know that recursion comes with its own set of trade-offs, and two of those are higher memory usage and a longer/taller stack that becomes increasingly more complex to debug.

Queueing-up work in a list (flattening the stack), then iteratively applying that work avoids both of those issues.

**Comment by cam72cam** [comment]
# RFC: Allowing template files to include other template files


## Use Case
In many scenarios, users wish to re-use pieces of templated configuration in tofu.  Some examples given by the community include instance startup scripts, cloud-init configs, and Azure Devops Templates.


## Why was this restriction added?
When the template_file provider was turned into a function, this restriction was added: https://github.com/opentofu/opentofu/commit/c753df6a933ef25197101664b1d7cbb5d115701d.  The only reasoning was to prevent infinite looping, resulting in a crash.  In the corresponding discussion in the terraform repo, reasoning was given that it was to prevent more complicated design patterns.


## How can we mitigate the reasoning behind the restriction?
I have created a PR that should mitigate the above concerns, #1250.  It keeps track of the depth of recursive calls using existing templatefile function context machinery.  When the maximum permissible depth (configurable) is reached, a error is returned that provides a clear stack trace of templates which led to the error.  This is compatible with all existing functions, including the newly added templatestring.

I also believe that we should let the users decide what design patterns they are willing to take on, instead of trying to force them in a given path.


## What are the risks in the current implementation?

### Memory
As @skyzyx points out, there is a risk of large templatefiles eating enough memory to cause a problem.  As this process is recursive (without the benefit of tail recursion) the full stack of read templatefiles need to be stored in memory until the stack can unwind.

I think that for most scenarios, this is a negligible concern.  A similar issue can be hit using many of the existing template language constructs like `${for}`.  Additionally, even if each call to templatefile were to take 1MB of memory (which is a lot of configuration), it would still only use 1GB of memory at its peak.

If my assumptions are wildly inaccurate, it is fairly trivial to add a separate tracker for estimated memory used by template data following the same code paths as the stack depth implementation.

### Time

If this feature is abused, it could slow down planning time as parsing templates is not a fast process.  Parsing a simple template 1k times takes up to 0.1s on my machine.  If truly complex setups start to emerge with templatefile, users could *theoretically* run into performance issues.

There does exist some low hanging fruit in the optimization path here, we parse the templatefile body every single invocation.  A cache could be added to the template loader to help in this admittedly unlikely scenario.


## Technical notes

Templatefile was already a bit of a special case of the language functions.  It needs to build an evaluation context that is specific to what's required in each template.  It is given a supplier of a list of functions registered with tofu.  This allows the templatefile function to be registered, and then later on use the full list of registered functions.

The original restriction modified the results of the function list to sub-out "templatefile" with a function that simply error'd.  By re-using that technique, we can sub-in a new instance of template function, with the stack depth tracker incremented by one.  If templatefile ever detects a stack depth greater than the allowed maximum, it returns an error.

That error is detected from calling instances of templatefile as the stack unwinds.  That "unwinding" is captured in the returned error and used to provide a nice user facing message.  That message is truncated in a best-effort manner to prevent flooding the user's console.  TF_LOG=debug provides a full stack trace if that ever proves necessary.

### Conceptual Example:
* templatefile("foo") -> `TemplateFile[1].load("foo").render("path", tofu_funcs + TemplateFile[2])`
  * templatefile("path") -> `TemplateFile[2].load("path").render("otherpath", tofu_funcs + TemplateFile[3])`
    * etc...
    
    
## Potential Alternate Implementations

@skyzyx points out that an iterative approach would mitigate a lot of the potential memory concerns and potentially make it easier to debug.  In a standard programming language, I would agree.  However I believe the fundamental concepts of how tofu handles evaluating templates prevents this approach from being feasible. We have no easy way of inspecting the payload for function calls ahead of time.  The only potential option that I see would be to attempt to render the template and inspect the resulting errors.

**Comment by cube2222** [comment]
LGTM @cam72cam!

Would be great to get this into 1.7 if it's not too much work, honestly.

**Comment by Yantrio** [comment]
LGTM @cam72cam.

Also I want to say thanks to @skyzyx for the input there too <3. Super appreciated!
